File: conductor/src/main/java/com/bluelinelabs/conductor/RouterTransaction.java
package com.bluelinelabs.conductor;

import android.os.Bundle;
import android.support.annotation.NonNull;

/**
 * Metadata used for adding {@link Controller}s to a {@link Router}.
 */
public class RouterTransaction {

    private static final String KEY_VIEW_CONTROLLER_BUNDLE = "RouterTransaction.controller.bundle";
    private static final String KEY_PUSH_TRANSITION = "RouterTransaction.pushControllerChangeHandler";
    private static final String KEY_POP_TRANSITION = "RouterTransaction.popControllerChangeHandler";
    private static final String KEY_TAG = "RouterTransaction.tag";
    private static final String KEY_ATTACHED_TO_ROUTER = "RouterTransaction.attachedToRouter";

    @NonNull final Controller controller;
    private String tag;

    private ControllerChangeHandler pushControllerChangeHandler;
    private ControllerChangeHandler popControllerChangeHandler;
    private boolean attachedToRouter;

    public static RouterTransaction with(@NonNull Controller controller) {
        return new RouterTransaction(controller);
    }

    private RouterTransaction(@NonNull Controller controller) {
        this.controller = controller;
    }

    RouterTransaction(@NonNull Bundle bundle) {
        controller = Controller.newInstance(bundle.getBundle(KEY_VIEW_CONTROLLER_BUNDLE));
        pushControllerChangeHandler = ControllerChangeHandler.fromBundle(bundle.getBundle(KEY_PUSH_TRANSITION));
        popControllerChangeHandler = ControllerChangeHandler.fromBundle(bundle.getBundle(KEY_POP_TRANSITION));
        tag = bundle.getString(KEY_TAG);
        attachedToRouter = bundle.getBoolean(KEY_ATTACHED_TO_ROUTER);
    }

    void onAttachedToRouter() {
        attachedToRouter = true;
    }

    public Controller controller() {
        return controller;
    }

    public String tag() {
        return tag;
    }

    public RouterTransaction tag(String tag) {
        if (!attachedToRouter) {
            this.tag = tag;
            return this;
        } else {
            throw new RuntimeException(getClass().getSimpleName() + "s can not be modified after being added to a Router.");
        }
    }

    public ControllerChangeHandler pushChangeHandler() {
        ControllerChangeHandler handler = controller.getOverriddenPushHandler();
        if (handler == null) {
            handler = pushControllerChangeHandler;
        }
        return handler;
    }

    public RouterTransaction pushChangeHandler(ControllerChangeHandler handler) {
        if (!attachedToRouter) {
            pushControllerChangeHandler = handler;
            return this;
        } else {
            throw new RuntimeException(getClass().getSimpleName() + "s can not be modified after being added to a Router.");
        }
    }

    public ControllerChangeHandler popChangeHandler() {
        ControllerChangeHandler handler = controller.getOverriddenPopHandler();
        if (handler == null) {
            handler = popControllerChangeHandler;
        }
        return handler;
    }

    public RouterTransaction popChangeHandler(ControllerChangeHandler handler) {
        if (!attachedToRouter) {
            popControllerChangeHandler = handler;
            return this;
        } else {
            throw new RuntimeException(getClass().getSimpleName() + "s can not be modified after being added to a Router.");
        }
    }

    /**
     * Used to serialize this transaction into a Bundle
     */
    public Bundle saveInstanceState() {
        Bundle bundle = new Bundle();

        bundle.putBundle(KEY_VIEW_CONTROLLER_BUNDLE, controller.saveInstanceState());

        if (pushControllerChangeHandler != null) {
            bundle.putBundle(KEY_PUSH_TRANSITION, pushControllerChangeHandler.toBundle());
        }
        if (popControllerChangeHandler != null) {
            bundle.putBundle(KEY_POP_TRANSITION, popControllerChangeHandler.toBundle());
        }

        bundle.putString(KEY_TAG, tag);
        bundle.putBoolean(KEY_ATTACHED_TO_ROUTER, attachedToRouter);

        return bundle;
    }

}

File: conductor/src/main/java/com/bluelinelabs/conductor/RouterTransaction.kt
package com.bluelinelabs.conductor

import android.os.Bundle
import androidx.annotation.RestrictTo
import androidx.annotation.RestrictTo.Scope.LIBRARY
import com.bluelinelabs.conductor.internal.TransactionIndexer

private const val INVALID_INDEX = -1
private const val KEY_VIEW_CONTROLLER_BUNDLE = "RouterTransaction.controller.bundle"
private const val KEY_PUSH_TRANSITION = "RouterTransaction.pushControllerChangeHandler"
private const val KEY_POP_TRANSITION = "RouterTransaction.popControllerChangeHandler"
private const val KEY_TAG = "RouterTransaction.tag"
private const val KEY_INDEX = "RouterTransaction.transactionIndex"
private const val KEY_ATTACHED_TO_ROUTER = "RouterTransaction.attachedToRouter"

/**
 * Metadata used for adding [Controller]s to a [Router].
 */
class RouterTransaction
private constructor(
  @get:JvmName("controller")
  val controller: Controller,
  private var tag: String? = null,
  private var pushControllerChangeHandler: ControllerChangeHandler? = null,
  private var popControllerChangeHandler: ControllerChangeHandler? = null,
  private var attachedToRouter: Boolean = false,
  @RestrictTo(LIBRARY)
  var transactionIndex: Int = INVALID_INDEX
) {


  @RestrictTo(LIBRARY)
  internal constructor(bundle: Bundle) : this(
    controller = Controller.newInstance(bundle.getBundle(KEY_VIEW_CONTROLLER_BUNDLE)!!),
    pushControllerChangeHandler = ControllerChangeHandler.fromBundle(
      bundle.getBundle(
        KEY_PUSH_TRANSITION
      )
    ),
    popControllerChangeHandler = ControllerChangeHandler.fromBundle(
      bundle.getBundle(
        KEY_POP_TRANSITION
      )
    ),
    tag = bundle.getString(KEY_TAG),
    transactionIndex = bundle.getInt(KEY_INDEX),
    attachedToRouter = bundle.getBoolean(KEY_ATTACHED_TO_ROUTER)
  )

  fun onAttachedToRouter() {
    attachedToRouter = true
  }

  fun tag(): String? = tag

  fun tag(tag: String?): RouterTransaction {
    return if (!attachedToRouter) {
      this.tag = tag
      this
    } else {
      throw RuntimeException(javaClass.simpleName + "s can not be modified after being added to a Router.")
    }
  }

  fun pushChangeHandler(): ControllerChangeHandler? {
    return controller.overriddenPushHandler ?: pushControllerChangeHandler
  }

  fun pushChangeHandler(handler: ControllerChangeHandler?): RouterTransaction {
    return if (!attachedToRouter) {
      pushControllerChangeHandler = handler
      this
    } else {
      throw RuntimeException("${javaClass.simpleName}s can not be modified after being added to a Router.")
    }
  }

  fun popChangeHandler(): ControllerChangeHandler? {
    return controller.overriddenPopHandler ?: popControllerChangeHandler
  }

  fun popChangeHandler(handler: ControllerChangeHandler?): RouterTransaction {
    return if (!attachedToRouter) {
      popControllerChangeHandler = handler
      this
    } else {
      throw RuntimeException("${javaClass.simpleName}s can not be modified after being added to a Router.")
    }
  }

  fun ensureValidIndex(indexer: TransactionIndexer) {
    if (transactionIndex == INVALID_INDEX) {
      transactionIndex = indexer.nextIndex()
    }
  }

  /**
   * Used to serialize this transaction into a Bundle
   */
  fun saveInstanceState(): Bundle = Bundle().apply {
    putBundle(KEY_VIEW_CONTROLLER_BUNDLE, controller.saveInstanceState())
    pushControllerChangeHandler?.let { putBundle(KEY_PUSH_TRANSITION, it.toBundle()) }
    popControllerChangeHandler?.let { putBundle(KEY_POP_TRANSITION, it.toBundle()) }
    putString(KEY_TAG, tag)
    putInt(KEY_INDEX, transactionIndex)
    putBoolean(KEY_ATTACHED_TO_ROUTER, attachedToRouter)
  }

  companion object {

    @JvmStatic
    fun with(controller: Controller): RouterTransaction = RouterTransaction(controller)
  }
}

fun Controller.asTransaction(
  popChangeHandler: ControllerChangeHandler? = null,
  pushChangeHandler: ControllerChangeHandler? = null
): RouterTransaction {
  return RouterTransaction.with(this)
    .pushChangeHandler(pushChangeHandler)
    .popChangeHandler(popChangeHandler)
}
